### PR TITLE
Implement custom function constraints

### PR DESCRIPTION
**Pull request by llambeau**

### CODE REVIEW

**Comment by llambeau** [comment]
Tiny problem here:

SubType calls `accept` on every constraint at dressing time (`#_mDress`) but also on `#_include`. In the second case, we don't have access to the world though.

File: specs/unit/constraint/accept_spec.coffee [hunk 1]
```diff
@@ -61,3 +61,29 @@ describe "Constraint#accept", ->
...
+    it 'accepts even numbers', ->
+      should(constraint.accept(2, world)).equal(true)
+      should(constraint.accept(4, world)).equal(true)
+      should(constraint.accept(10, world)).equal(true)
+
+    it 'rejects odd numbers', ->
+      should(constraint.accept(9, world)).equal(false)
+      should(constraint.accept(1, world)).equal(false)
+
+  describe 'with a dotted function', ->
```
Comment (blambeau) [inline specs/unit/constraint/accept_spec.coffee]: with a qualified function name

File: src/finitio/parser.pegjs [hunk 2]
```diff
@@ -92,6 +92,9 @@ constraint =
 / spacing '::' spacing rx:range_literal {
     return [{ range: rx }];
   }
+/ spacing '::' spacing rx:function_literal {
```
Comment (blambeau) [inline src/finitio/parser.pegjs]: rx => fn

File: src/finitio/support/constraint.coffee [hunk 3]
```diff
@@ -13,7 +13,7 @@ class Constraint
   isAnonymous: ->
     !@name?
 
-  accept: (arg) ->
+  accept: (arg, world) ->
     throw new Error("Constraint is an abstract class")
```
Comment (blambeau) [inline src/finitio/support/constraint.coffee]: On second thoughts, I don't agree with this, but making something better requires some refactoring. Let keep it for now.

File: src/finitio/support/constraint.coffee [hunk 4]
```diff
@@ -40,13 +40,33 @@ class Constraint.Native extends Constraint
...
+    func = world
+
+    $u.each identifiers, (id, idx) ->
+      path += '.' if idx > 0
+      path += id
+      func = func[id]
+      unless func
+        throw new Error("#{path} is undefined");
+
+    func(arg)
```
Comment (blambeau) [inline src/finitio/support/constraint.coffee]: Please refactor towards a functional programming style, e.g. folr/reduce or something like that

File: src/finitio/support/constraint.coffee [hunk 5]
```diff
@@ -40,13 +40,33 @@ class Constraint.Native extends Constraint
...
+
+  nativeToString: () ->
+    '&' + @native
+
```
Comment (blambeau) [inline src/finitio/support/constraint.coffee]: Add implementation for `equals`

**Pull request merged**